### PR TITLE
OKTA-363616:  statemanager.renew

### PR DESCRIPTION
**Pull request by bryanapellanes-okta**
- Implemented statemanager.renew
- Implemented statemanager.introspect
- Added tests

### CODE REVIEW

File: Okta.Xamarin/Okta.Xamarin/IOidcClient.cs [hunk 1]
```diff
@@ -27,19 +35,78 @@ public interface IOidcClient
...
+        /// <summary>
+        /// Gets information about the state of the specified token.
+        /// </summary>
+        /// <param name="tokenType">The type of the token to introspect.</param>
+        /// <param name="token">The token to introspect.</param>
+        /// <param name="authorizationServerId">The authorization server ID.</param>
+        /// <returns>Dictionary{string, object}.</returns>
+        Task<Dictionary<string, object>> IntrospectAsync(TokenKind tokenType, string token, string authorizationServerId = "default");
+
+		Task<T> RenewAsync<T>(string refreshToken, bool refreshIdToken = false, string authorizationServerId = "default");
```
Comment (bryanapellanes-okta) [inline Okta.Xamarin/Okta.Xamarin/IOidcClient.cs:110]: I'll fix formatting and add doc in upcoming PR.

File: Okta.Xamarin/Okta.Xamarin/Models/Serializable.cs [hunk 2]
```diff
@@ -31,6 +32,11 @@ public string ToJson()
             return JsonConvert.SerializeObject(this);
         }
 
+		public Dictionary<string, object> ToDictionary()
+		{
+			return JsonConvert.DeserializeObject<Dictionary<string, object>>(ToJson());
+		}
```
Comment (bryanapellanes-okta) [inline Okta.Xamarin/Okta.Xamarin/Models/Serializable.cs:38]: I'll fix formatting in next PR

File: Okta.Xamarin/Okta.Xamarin/OktaContext.cs [hunk 3]
```diff
@@ -126,9 +163,45 @@ public static void AddSignOutCompletedListener(EventHandler<SignOutEventArgs> si
...
+        /// Convenience method to add a listener to the OktaContext.Current.IntrospectStarted event.
+        /// </summary>
+        /// <param name="introspectEventHandler">The event handler.</param>
+        public static void AddIntrospectStartedListener(EventHandler<IntrospectEventArgs> introspectEventHandler)
+        {
+            Current.IntrospectStarted += introspectEventHandler;
+        }
+
+        /// <summary>
+        /// Convenience method to add a listene to the OktaContenxt.Current.InstrospectCompleted event.
```
Comment (andriizhegurov-okta) [inline Okta.Xamarin/Okta.Xamarin/OktaContext.cs:199]:
```suggestion
        /// Convenience method to add a listener to the OktaContenxt.Current.InstrospectCompleted event.
```

File: Okta.Xamarin/Okta.Xamarin/OktaContext.cs [hunk 4]
```diff
@@ -185,14 +258,66 @@ public virtual async Task SignOutAsync(IOidcClient oidcClient = null)
         /// </summary>
         /// <param name="tokenType">The type of token to revoke</param>
         /// <returns>Task</returns>
-        public virtual async Task RevokeTokenAsync(TokenType tokenType)
+        public virtual async Task RevokeTokenAsync(TokenKind tokenType)
```
Comment (andriizhegurov-okta) [inline Okta.Xamarin/Okta.Xamarin/OktaContext.cs:261]: It's probably worth renaming the parameter as well: `tokenType` -> `tokenKind`

Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/OktaContext.cs:261]: Good catch.  I'll update shortly.

File: Okta.Xamarin/Okta.Xamarin/OktaXamarinUserAgent.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+// </copyright>
+
+using System;
+using System.Reflection;
+
+namespace Okta.Xamarin
+{
+	public class OktaXamarinUserAgent
+    {
+        public static string Value => new Lazy<string>(() => $"Okta-Xamarin-Sdk/{Assembly.GetExecutingAssembly().GetName().Version}").Value;
```
Comment (andriizhegurov-okta) [inline Okta.Xamarin/Okta.Xamarin/OktaXamarinUserAgent.cs:13]: Also this needs to be changed the same as in PR #35

Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/OktaXamarinUserAgent.cs:13]: That change will be picked up when 35 is merged.

Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/OktaXamarinUserAgent.cs:13]: Or I could change it here and just merge this PR instead of 35.  Small enough that it shouldn't be a problem either way.

File: Okta.Xamarin/Okta.Xamarin/ContentPageExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.
+// </copyright>
+
+using System.Collections.Generic;
+using Xamarin.Forms;
+
+namespace Okta.Xamarin
+{
+    /// <summary>
+    /// Extension metods for ContentPage.
```
Comment (andriizhegurov-okta) [inline Okta.Xamarin/Okta.Xamarin/ContentPageExtensions.cs:12]:
```suggestion
    /// Extension methods for ContentPage.
```

File: Okta.Xamarin/Okta.Xamarin/IOidcClient.cs [hunk 7]
```diff
@@ -57,7 +63,7 @@ public interface IOidcClient
         /// <param name="accessToken">The access token used for authorization.</param>
```
Comment (andriizhegurov-okta) [inline Okta.Xamarin/Okta.Xamarin/IOidcClient.cs:63]:
```suggestion
```

File: Okta.Xamarin/Okta.Xamarin/IntrospectEventArgs.cs [hunk 8]
```diff
@@ -20,11 +20,16 @@ public class IntrospectEventArgs : EventArgs
         /// <summary>
         /// Gets or sets the token type.
```
Comment (andriizhegurov-okta) [inline Okta.Xamarin/Okta.Xamarin/IntrospectEventArgs.cs:21]:
```suggestion
        /// Gets or sets the token kind.
```

File: Okta.Xamarin/Okta.Xamarin/IntrospectOptions.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+            AuthorizationServerId = "default";
+        }
+
+        /// <summary>
+        /// Gets or sets the token that is the target of introspection.
+        /// </summary>
+        public string Token { get; set; }
+
+        /// <summary>
+        /// Gets or sets the type of the target token.
```
Comment (andriizhegurov-okta) [inline Okta.Xamarin/Okta.Xamarin/IntrospectOptions.cs:27]:
```suggestion
        /// Gets or sets the kind of the target token.
```

File: Okta.Xamarin/Okta.Xamarin/OidcClient.cs [hunk 10]
```diff
@@ -113,6 +117,45 @@ public async Task<IOktaStateManager> AuthenticateAsync(string sessionToken)
...
+        /// <returns>Dicationary{string, object}.</returns>
+        public async Task<Dictionary<string, object>> IntrospectAsync(IntrospectOptions options)
+        {
+            return await IntrospectAsync(options.TokenKind, options.Token, options.AuthorizationServerId);
+        }
+
+        /// <summary>
+        /// Gets introspection details.
+        /// </summary>
+        /// <param name="accessToken">The access token used to authroize the request.</param>
```
Comment (andriizhegurov-okta) [inline Okta.Xamarin/Okta.Xamarin/OidcClient.cs:133]:
```suggestion
```

File: Okta.Xamarin/Okta.Xamarin/OktaContext.cs [hunk 11]
```diff
@@ -186,14 +214,32 @@ public static void AddIntrospectStartedListener(EventHandler<IntrospectEventArgs
...
+        /// Convenience method to add a listener to the OktaContext.Current.RenewStarted event.
+        /// </summary>
+        /// <param name="renewEventHandler">The event handler.</param>
+        public static void AddRenewStartedListener(EventHandler<RenewEventArgs> renewEventHandler)
+        {
+            Current.RenewStarted += renewEventHandler;
+        }
+
+        /// <summary>
+        /// Convenience method to add a listener to the OktaContext.Current.RenewStarted event.
```
Comment (andriizhegurov-okta) [inline Okta.Xamarin/Okta.Xamarin/OktaContext.cs:235]:
```suggestion
        /// Convenience method to add a listener to the OktaContext.Current.RenewCompleted event.
```

File: Okta.Xamarin/Okta.Xamarin/OktaContext.cs [hunk 12]
```diff
@@ -244,23 +290,54 @@ public virtual async Task SignOutAsync(IOidcClient oidcClient = null)
         }
 
         /// <summary>
-        /// Revoke token of the specified type.
+        /// Revoke token of the specified kind.
+        /// </summary>
+        /// <param name="tokenKind">The type of token to revoke.</param>
```
Comment (andriizhegurov-okta) [inline Okta.Xamarin/Okta.Xamarin/OktaContext.cs:295]:
```suggestion
        /// <param name="tokenKind">The kind of token to revoke.</param>
```

File: Okta.Xamarin/Okta.Xamarin/RenewOptions.cs [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+        /// <summary>
+        /// Gets or sets the refresh token.
+        /// </summary>
+        public string RefreshToken { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the id token is refreshed.
+        /// </summary>
+        public bool RefreshIdToken { get; set; }
```
Comment (andriizhegurov-okta) [inline Okta.Xamarin/Okta.Xamarin/RenewOptions.cs:29]: May be `IsTokenRefreshed` is a better name? (just like `IsAuthenticated` and other boolean props)

File: Okta.Xamarin/Okta.Xamarin/ViewModels/DiagnosticsViewModel.cs [hunk 14]
```diff
@@ -3,36 +3,81 @@
...
+                this.ClaimsPrincipal = args.UserInfo as ClaimsPrincipal; // user info may be Dictionary<string, object> or ClaimsPrincipal or generic type; will be null if incorrect 'as' cast
+            };
+            OktaContext.Current.IntrospectCompleted += (sender, args) =>
+            {
+                this.Page.DisplayData(args.Response as Dictionary<string, object>, "Introspection");
+            };
+            OktaContext.Current.RenewCompleted += (sender, args) =>
+            {
+                this.Page.DisplayData(args.Response.ToDictionary(), "Renewal");
+            };
```
Comment (andriizhegurov-okta) [inline Okta.Xamarin/Okta.Xamarin/ViewModels/DiagnosticsViewModel.cs:41]: Just thoughts:
What about introducing functions in the View for populating Introspection and Renewal responses and calling them then from here like:
 `this.Page.DisplayIntrospectionData(args.Response as Dictionary<string, object>)`
...
`this.Page.DisplayRenewalData(args.Response.ToDictionary());`

and those functions will be calling `.DisplayData` extension method if needed. So private sections names will not be mentioned outside the view.

File: Okta.Xamarin/Okta.Xamarin/RenewOptions.cs (hunk 13 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/RenewOptions.cs:29]: This parameter affects the behavior of the renew operation (as in hey renew operation please `RefreshIdToken` also).

File: Okta.Xamarin/Okta.Xamarin/ViewModels/DiagnosticsViewModel.cs (hunk 14 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/ViewModels/DiagnosticsViewModel.cs:41]: I have additional items coming in v1.1 that will address that concern.

**Review by andriizhegurov-okta — APPROVED** [review]
👍

File: Okta.Xamarin/Okta.Xamarin.Test/TestOktaStateManager.cs [hunk 15]
```diff
@@ -19,7 +19,7 @@ public TestOktaStateManager(string accessToken, string idToken = null, string re
 
         public int RevokeAsyncCallCount { get; set; }
 
-        public override Task RevokeAsync(TokenType tokenType)
+        public override Task RevokeAsync(TokenKind tokenType)
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin.Test/TestOktaStateManager.cs:22]: Nit: Rename var name from tokenType to tokenKind for consistency.

File: Okta.Xamarin/Okta.Xamarin/IOktaStateManager.cs [hunk 16]
```diff
@@ -16,38 +16,125 @@ namespace Okta.Xamarin
...
 
+        /// <summary>
+        /// Gets or sets the client.
+        /// </summary>
+        IOidcClient Client { get; set; }
+
+        /// <summary>
+        /// Gets the expiration time.
+        /// </summary>
         DateTime? Expires { get; }
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/IOktaStateManager.cs:37]: For expiration, we should use [`DateTimeOffset?`](https://docs.microsoft.com/en-us/dotnet/standard/datetime/choosing-between-datetime#the-datetimeoffset-structure) instead.

File: Okta.Xamarin/Okta.Xamarin/IOktaStateManager.cs [hunk 17]
```diff
@@ -16,38 +16,125 @@ namespace Okta.Xamarin
...
 
+        /// <summary>
+        /// Gets the id token.
+        /// </summary>
         string IdToken { get; }
 
+        /// <summary>
+        /// Gets a value indicating whether or not there is a current non-expired <see cref="AccessToken"/>, indicating the user is currently successfully authenticated.
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/IOktaStateManager.cs:45]: I think we should rephrase this to avoid "indicating" twice.

File: Okta.Xamarin/Okta.Xamarin/IOktaStateManager.cs [hunk 18]
```diff
@@ -16,38 +16,125 @@ namespace Okta.Xamarin
...
         string RefreshToken { get; }
 
+        /// <summary>
+        /// Gets the scope.
+        /// </summary>
         string Scope { get; }
 
+        /// <summary>
+        /// Gets the type of the AccessToken, for example "Bearer".
+        /// </summary>
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/IOktaStateManager.cs:66]: What's the difference between TokenType vs TokenKind? I noticed we have 2 diff ways to categorize tokens.

File: Okta.Xamarin/Okta.Xamarin/IOktaStateManager.cs [hunk 19]
```diff
@@ -16,38 +16,125 @@ namespace Okta.Xamarin
...
-
-		Task<Dictionary<string, object>> GetUserAsync(string authorizationServerId = "default");
-
-		Task<ClaimsPrincipal> GetClaimsPrincipalAsync(string authorizationServerId = "default");
-
-        Task<string> RenewAsync();
+        /// <summary>
+        /// Gets the token of the specified type.
+        /// </summary>
+        /// <param name="tokenKind">The token type.</param>
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/IOktaStateManager.cs:77]: Nit: Should we use kind instead type in the comments?

File: Okta.Xamarin/Okta.Xamarin/IntrospectOptions.cs [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+
+        /// <summary>
+        /// Gets or sets the kind of the target token.
+        /// </summary>
+        public TokenKind TokenKind{ get; set; } 
+
+        /// <summary>
+        /// Gets or sets the authorization server id.
+        /// </summary>
+        public string AuthorizationServerId{ get; set; }
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/IntrospectOptions.cs:34]: Nit: Leave a space between the prop and {}

File: Okta.Xamarin/Okta.Xamarin/OktaConfig.cs [hunk 21]
```diff
@@ -154,33 +154,38 @@ public static async Task<OktaConfig> LoadFromJsonFileAsync(string filename)
...
                 return $"{OktaDomain}/oauth2/{AuthorizationServerId}/v1/authorize";
             }
+
             return AuthorizeUri;
         }
 
+        /// <summary>
+        /// Gets the logout url used for loggin a user out.
+        /// </summary>
+        /// <returns>The computed logout url used for logging out.</returns>
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/OktaConfig.cs:171]: NIt: URL

File: Okta.Xamarin/Okta.Xamarin/OktaStateManager.cs [hunk 22]
```diff
@@ -28,12 +26,12 @@ public OktaStateManager()
...
-        /// <param name="refreshToken"></param>
-        /// <param name="expiresIn"></param>
-        /// <param name="scope"></param>
+        /// <param name="accessToken">The access token.</param>
+        /// <param name="tokenType">The token type.</param>
+        /// <param name="idToken">The ID token.</param>
+        /// <param name="refreshToken">The refresh token.</param>
+        /// <param name="expiresIn">Expires in.</param>
+        /// <param name="scope">Scope.</param>
         public OktaStateManager(string accessToken, string tokenType, string idToken = null, string refreshToken = null, int? expiresIn = null, string scope = null)
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/OktaStateManager.cs:35]: Same here, tokenType vs tokenKind. SHould we use just one name for consistency?

File: Okta.Xamarin/Okta.Xamarin/RenewResponse.cs [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+        /// Gets or sets the token type.
+        /// </summary>
+        [JsonProperty("token_type")]
+        public string TokenType { get; set; }
+
+        /// <summary>
+        /// Gets or sets the expires in.
+        /// </summary>
+        [JsonProperty("expires_in")]
+        public int ExpiresIn { get; set; }
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/RenewResponse.cs:26]: Should we specify the time unit in the comments (i.e seconds)?

File: Okta.Xamarin/Okta.Xamarin/ViewModels/RevokeRefreshTokenCommand.cs [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+    /// Command used to revoke the refresh token.
+    /// </summary>
+    public class RevokeRefreshTokenCommand : Command
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RevokeRefreshTokenCommand"/> class.
+        /// </summary>
+        public RevokeRefreshTokenCommand()
+            : base(async () => await OktaContext.Current.RevokeAsync(TokenKind.RefreshToken))
+        {
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/ViewModels/RevokeRefreshTokenCommand.cs:18]: Nit: FIx indentation.

File: README.md [hunk 25]
```diff
@@ -291,60 +297,336 @@ To receive a refresh token along with id and access tokens do the following:
...
 ### OktaContext.Current
 
-The `OktaContext.Current` singleton provides a top level entry point into Okta functionality.  To begin the login flow use the `SignIn` method:
+The `OktaContext.Current` singleton provides a top level entry point into Okta functionality.  
+
+#### SignInAsync
+
+Signs a user in. See also, [SignInStarted event](#signinstarted-event) and [SignInCompleted event](#signincompleted-event).
+```csharp
+await OktaContext.Current.SignInAsync();
```
Comment (laura-rodriguez) [inline README.md:306]: Should we also add details of what this method returns?

**Review by laura-rodriguez — APPROVED** [review]
Minor suggestions, then LGTM.

File: Okta.Xamarin/Okta.Xamarin/IOktaStateManager.cs (hunk 18 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/IOktaStateManager.cs:66]: TokenType is what the server API returns describing the token for use as an `Authorization` header; the only value I've seen so far is "Bearer" but I assume other valid values would include "SSWS" or "basic".  To differentiate between the kind of token used in an `Authorization` header and the tokens that a developer will have related to the user's authentication state I created the TokenKind enum; valid values are `IdToken`, `AccessToken` and `RefreshToken`.  Because these are categorically different I wanted to disambiguate.

File: Okta.Xamarin/Okta.Xamarin/OktaStateManager.cs (hunk 22 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/OktaStateManager.cs:35]: This one actually is the tokenType value returned by the server.  For now it is recorded in the StateManager to advise what would be the appropriate Authorization header value prefix (see previous comment on this topic); this is my reasoning thus far.  Usage will reveal whether this is useful.

File: Okta.Xamarin/Okta.Xamarin/RenewResponse.cs (hunk 23 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/RenewResponse.cs:26]: We probably should; it wasn't clear to me what the actual unit was.  I'll add that.
